src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,462 @@
+#define CASE(X) \
+    case RoomAlignEnum::X: \
+        return #X;
+        switch (e) {
+            X_FOREACH_RoomAlignEnum(CASE)
+        }
+#undef CASE
+        assert(false); // reachable only on invalid RoomAlignEnum
+    }
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::doorFlagName(const DoorFlagEnum e)
+{
+#define CASE(UPPER_CASE, lower_case, CamelCase, friendly) \
+    case DoorFlagEnum::UPPER_CASE: \
+        return #UPPER_CASE;
+    switch (e) {
+        X_FOREACH_DOOR_FLAG(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid DoorFlagEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::exitFlagName(const ExitFlagEnum e)
+{
+#define CASE(UPPER_CASE, lower_case, CamelCase, friendly) \
+    case ExitFlagEnum::UPPER_CASE: \
+        return #UPPER_CASE;
+    switch (e) {
+        X_FOREACH_EXIT_FLAG(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid DoorFlagEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::lightName(const RoomLightEnum e)
+{
+    if (e != RoomLightEnum::UNDEFINED) {
+#define CASE(X) \
+    case RoomLightEnum::X: \
+        return #X;
+        switch (e) {
+            X_FOREACH_RoomLightEnum(CASE)
+        }
+#undef CASE
+        assert(false); // reachable only on invalid RoomLightEnum
+    }
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::loadFlagName(const RoomLoadFlagEnum e)
+{
+#define CASE(X) \
+    case RoomLoadFlagEnum::X: \
+        return #X;
+    switch (e) {
+        X_FOREACH_ROOM_LOAD_FLAG(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid RoomLoadFlagEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::markClassName(const InfoMarkClassEnum e)
+{
+#define CASE(X) \
+    case InfoMarkClassEnum::X: \
+        return #X;
+    switch (e) {
+        X_FOREACH_INFOMARK_CLASS(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid InfoMarkClassEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::markTypeName(const InfoMarkTypeEnum e)
+{
+#define CASE(X) \
+    case InfoMarkTypeEnum::X: \
+        return #X;
+    switch (e) {
+        X_FOREACH_INFOMARK_TYPE(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid InfoMarkTypeEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::mobFlagName(const RoomMobFlagEnum e)
+{
+#define CASE(X) \
+    case RoomMobFlagEnum::X: \
+        return #X;
+    switch (e) {
+        X_FOREACH_ROOM_MOB_FLAG(CASE)
+    }
+#undef CASE
+    assert(false); // reachable only on invalid RoomMobFlagEnum
+    return "";
+}
+
+NODISCARD const char *XmlMapStorage::portableName(const RoomPortableEnum e)
+{
+    if (e != RoomPortableEnum::UNDEFINED) {
+#define CASE(X) \
+    case RoomPortableEnum::X: \
+        return #X;
+        switch (e) {
+            X_FOREACH_RoomPortableEnum(CASE)
+        }
+#undef CASE
+        assert(false); // reachable only on invalid RoomPortableEnum
+    }
+    return "";
```

Comment:
Doesn't this mean we're going to persist an empty string to the file if it's undefined? Shouldn't we just omit the element entirely?

Reply:
We omit the element entirely.
The empty string is detected in `XmlMapStorage::saveXmlElement` and `XmlMapStorage::saveXmlAttribute`
which omit the entire element (or attribute) if the value to save is the empty string.

Reply:
How about using std::optional instead of an empty string? It's more clear then how it will be used

Reply:
If you want, ok.
I thought the empty string was already enough a special value.

Reply:
Looking better, the lines
```C++
    saveXmlElement(stream, "description", room.getDescription().toQString());
    saveXmlElement(stream, "contents", room.getContents().toQString());
    saveXmlElement(stream, "note", room.getNote().toQString());
```
would become more complicated, because for those we want to treat these empty string as "do not save the element", which the current `XmlMapStorage::saveXmlElement` does automatically.

If `XmlMapStorage::saveXmlElement` and `XmlMapStorage::saveXmlAttribute` are changed to accept `std::optional<QString>` with the semantics that only `std::nullopt` omits the element entirely,
then the code above must check for empty TaggedString, i.e.

```C++
    if (!room.getDescription().isEmpty()) {
        saveXmlElement(stream, "description", room.getDescription().toQString());
    }
    if (!room.getContents().isEmpty()) {
        saveXmlElement(stream, "contents", room.getContents().toQString());
    }
    if (!room.getNote().isEmpty()) {
        saveXmlElement(stream, "note", room.getNote().toQString());
    }
```
which defeats the very purpose of having `saveXmlElement()` accept a `std::optional<QString>`

In other words, the `.mm2xml` format already treats all empty strings specially - by **not** storing the element or attribute containing only them (or **should** treat empty strings in this way: I may have missed some cases, which in my opinition should be fixed)
so I do not see much need to introduce a different special value - an empty `std::optional<QString>` - to indicate the same concept.

Reply:
Agreed, seems like the best solution then. Thanks for digging into it 👍